doc/source/reference/c-api/types-and-structures.rst

```diff
@@ -290,7 +290,7 @@ PyArrayDescr_Type and PyArray_Descr
           npy_intp alignment;
           NpyAuxData *c_metadata;
           npy_hash_t hash;
-          void *reserved_null;  // unused field, must be NULL.
+          void *reserved_null[2];  // unused field, must be NULLed.
```

Comment:
I don't understand why this is needed. In any case we now expect users to allocate via an API call, so we can arbitrarily add fields when needed in the future.

Reply:
We expect downstream to subclass dtype, which means adding additional fields (e.g. the string dtype adds fields for its `allocator` and `na_object`).

So doing that would be painful if we don't promise the struct size to be stable (I am fine for that with `ndarray`, since I don't really expect subclassing, but here we do).
OTOH, we always have different mechanisms (through new accessor methods or with slot offsets like Python does it for `vectorcall`) to add new fields.

So I am OK either way, just to explain the reasoning.

Reply:
Ahh, subclassing makes sense, thanks for explaining.